redis/commands/mset.cc

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright (C) 2019 ScyllaDB
+ */
+
+/*
+ * This file is part of Scylla.
+ *
+ * Scylla is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "redis/commands/mset.hh"
+#include "seastar/core/shared_ptr.hh"
+#include "redis/request.hh"
+#include "redis/reply.hh"
+#include "service_permit.hh"
+#include "types.hh"
+#include "service/storage_proxy.hh"
+#include "service/client_state.hh"
+#include "redis/options.hh"
+#include "redis/mutation_utils.hh"
+
+namespace redis {
+
+namespace commands {
+
+shared_ptr<abstract_command> mset::prepare(service::storage_proxy& proxy, request&& req)
+{
+    std::map<bytes, bytes> data_map;
+    if (req.arguments_size() == 0 || req.arguments_size() % 2 != 0) {
+        throw wrong_number_of_arguments_exception(req._command);
+    }
+
+    for (unsigned i = 0; i < req.arguments_size() / 2; i++) {
+        data_map.insert(std::make_pair(req._args[2 * i], req._args[2 * i + 1]));
+    }
+
+    return seastar::make_shared<mset> (std::move(req._command), std::move(data_map), 0);
+}
+
+future<redis_message> mset::execute(service::storage_proxy& proxy, redis::redis_options& options, service_permit permit)
```

Comment:
A question not directly related to this patch but worth thinking about:
Why do we have this separation between the prepare() and execute() steps? Why can't we have a single function which takes a request and both parses it and runs the operation? It can remove a lot of boilerplate code from these operations.... I think it can remove about 2/3rds of the code of each operation...

Reply:
The purpose is splitting the disunity argument parse, each command has special argument, and parse logic.

command_factory->create() is used to parse the cmd itself.
command_factory->prepare() is used to parse the cmd arguments
abstract_command->execute() is used to really execute the cmd

If we merge two functions together, it will convert Object-Oriented to procedure-oriented.

@fastio what's your opinion?

Reply:
It's not really related to this patch, so we don't need to decide now. It's just a matter of style so I'll leave it to you guys to decide which style you like best.